### PR TITLE
Java 11 compile

### PR DESCRIPTION
**Pull request by F43nd1r**
Started work on #2779 as noted in https://github.com/querydsl/querydsl/pull/2794#issuecomment-830772378

Main points:
- Replace org.reflections:reflections with io.github.classgraph:classgraph, because the old library isn't updated anymore.
- Tests were expecting the old `javax.annotation.Generated`, generify.
- Update several dependency versions, in case of ECJ that includes a group change.

### CODE REVIEW

**Comment by F43nd1r** [comment]
I'm trying to fix more of the tests, but I'm unsure on how to run them locally. The vagrant setup mentioned in the readme doesn't seem to work anymore. Is there a method/setup I can use?

**Comment by F43nd1r** [comment]
@jwgmeligmeyling I'm no scala expert. Accessing `scala.Enumeration$Value` in generated sources works with jdk 8, but not with 11.

```
Running com.querydsl.scala.GenericExporterTest
target/gen1/com/querydsl/scala/ext/QSomeEntity.scala:24: error: type Enumeration$Value is not a member of package scala
  val someScalaEnum = createComparable[scala.Enumeration$Value]("someScalaEnum")
                                             ^
Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 6.777 sec <<< FAILURE! - in com.querydsl.scala.GenericExporterTest
Export(com.querydsl.scala.GenericExporterTest)  Time elapsed: 6.776 sec  <<< FAILURE!
java.lang.AssertionError: Compilation failed
	at com.querydsl.scala.CompileTestUtils$.assertCompileSuccess(CompileTestUtils.scala:36)
	at com.querydsl.scala.CompileTestUtils$.assertCompileSuccess(CompileTestUtils.scala:27)
	at com.querydsl.scala.GenericExporterTest.Export(GenericExporterTest.scala:20)
```

**Comment by lpandzic** [comment]
The version defined in [scala module pom.xml](https://github.com/querydsl/querydsl/blob/master/querydsl-scala/pom.xml#L22) [doesn't support jdk 11](https://docs.scala-lang.org/overviews/jdk-compatibility/overview.html).

**Comment by lpandzic** [comment]
Try upgrading to 2.11.2

**Comment by F43nd1r** [comment]
@lpandzic this PR used `2.12.12` even before I started working on it. I'll try the downgrade anyway.

**Comment by F43nd1r** [comment]
Works for now. 
Long term this dependency should be upgraded to a version not 5+ years old, but I guess that's valid for most dependencies in here.

**Comment by lpandzic** [comment]
> 
> 
> Works for now.
> Long term this dependency should be upgraded to a version not 5+ years old, but I guess that's valid for most dependencies in here.

I agree, but the focus now should be getting 5.0.0 released.

File: querydsl-codegen-utils/src/main/java/com/querydsl/codegen/utils/SimpleCompiler.java [hunk 1]
```diff
@@ -47,36 +47,8 @@ protected static boolean isSureFireBooter(URLClassLoader cl) {
...
-                }
-            }
-            return paths.stream().collect(Collectors.joining(File.pathSeparator));
-        } catch (UnsupportedEncodingException e) {
-            throw new CodegenException(e);
-        } catch (IOException e) {
-            throw new CodegenException(e);
-        }
+    public static String getClassPath(ClassLoader cl) {
+        return new ClassGraph().overrideClassLoaders(cl).getClasspath();
```
Comment (jwgmeligmeyling) [inline querydsl-codegen-utils/src/main/java/com/querydsl/codegen/utils/SimpleCompiler.java:51]: A key design decision in QueryDSL 5 was keeping the transitive dependencies to a bare minimum. Some folks in the issue tracker mentioned even that frameworks should impose no further dependencies at all.

I know this dependency is just used in code gen and won't leak into the users application in _most_ cases. That is, unless the wrong dependency scope is used for the apt dependencies or a module with a runtime dependency on codegen is used (like `Alias.*`).

Was it necessary to update this code for Java 11?

**Review by jwgmeligmeyling — APPROVED** [review]

File: querydsl-codegen-utils/src/main/java/com/querydsl/codegen/utils/SimpleCompiler.java (hunk 1 above)
Comment (F43nd1r) [reply querydsl-codegen-utils/src/main/java/com/querydsl/codegen/utils/SimpleCompiler.java:51]: Classloaders aren't `URLClassLoader`s anymore. 

Alternative would be to copy/paste from that library (or any other source), but updating is necessary.

Comment (jwgmeligmeyling) [reply querydsl-codegen-utils/src/main/java/com/querydsl/codegen/utils/SimpleCompiler.java:51]: Oh well, the code relied on a dependency anyway, that we now got rid of 😄

File: querydsl-core/pom.xml [hunk 2]
```diff
@@ -98,4 +98,21 @@
       </plugin>
     </plugins>
```
Comment (jwgmeligmeyling) [inline querydsl-core/pom.xml:99]: Can you verify that the introduced dependency for `classgraph` is in `provided` scope here? (as it is brought transitively with the dependency on codegen)

File: src/main/resources/querydsl_checks_base.xml [hunk 3]
```diff
@@ -72,7 +69,7 @@
     <module name="WhitespaceAround"/>
 
     <module name="ModifierOrder"/>
-    <module name="RedundantModifier"/>
+<!--    <module name="RedundantModifier"/>-->
```
Comment (jwgmeligmeyling) [inline src/main/resources/querydsl_checks_base.xml:73]: Ah yes, I see I still had to reapply a few rules here. Somehow they started giving issues after the update.

Should we first merge this and then update the other PR?

File: querydsl-core/pom.xml (hunk 2 above)
Comment (F43nd1r) [reply querydsl-core/pom.xml:99]: Did you intend to comment somewhere else? Core doesn't depend on codegen

File: src/main/resources/querydsl_checks_base.xml (hunk 3 above)
Comment (F43nd1r) [reply src/main/resources/querydsl_checks_base.xml:73]: Yes, this first, the other will be simpler to resolve

File: querydsl-core/pom.xml (hunk 2 above)
Comment (F43nd1r) [reply querydsl-core/pom.xml:99]: classgraph scope is `test` for all modules except codegen and codegen-utils

**Comment by jwgmeligmeyling** [comment]
I noticed that coverage report on Java 11 gives issues on the master branch: https://github.com/querydsl/querydsl/runs/2729469140 . It is not enabled on the other branches, hence the issue.

We might want to look into it some day.

**Pull request merged**